File: Deber 2/ejercicio_4/fwgc_solver.pl
% fwgc_solver.pl - Solver del problema Farmer, Wolf, Goat, Cabbage en SWI-Prolog
% Representación del estado: state(F, W, G, C) con lados l (izquierda) y r (derecha)

:- initialization(run, main).

% Estado inicial y objetivo
start(state(l,l,l,l)).
goal(state(r,r,r,r)).

% Lados opuestos
opp(l, r).
opp(r, l).

% Un estado es seguro si no se deja Lobo con Cabra sin el Granjero,
% ni Cabra con Col sin el Granjero
safe(state(F,W,G,C)) :-
    \+ (W = G, F \= W),
    \+ (G = C, F \= G).

% Movimientos posibles (el granjero siempre cruza):
% - Solo
move(state(F,W,G,C), state(F1,W,G,C), farmer) :-
    opp(F, F1),
    safe(state(F1,W,G,C)).
% - Con el lobo
move(state(F,W,G,C), state(F1,W1,G,C), farmer_wolf) :-
    F = W, opp(F, F1), opp(W, W1),
    safe(state(F1,W1,G,C)).
% - Con la cabra
move(state(F,W,G,C), state(F1,W,G1,C), farmer_goat) :-
    F = G, opp(F, F1), opp(G, G1),
    safe(state(F1,W,G1,C)).
% - Con la col
move(state(F,W,G,C), state(F1,W,G,C1), farmer_cabbage) :-
    F = C, opp(F, F1), opp(C, C1),
    safe(state(F1,W,G,C1)).

% Envoltura sin acción para facilitar el uso con dfs/bfs
move(State, Next) :- move(State, Next, _).

% --- Backtracking (DFS) ---
% Encuentra una ruta cualquiera desde start a goal (no garantiza mínimos pasos)
solve(Path) :-
    start(S), goal(G),
    dfs(S, G, [S], Path).

dfs(State, Goal, _, [State]) :-
    State = Goal, !.
dfs(State, Goal, Visited, [State|Path]) :-
    move(State, Next),
    \+ member(Next, Visited),
    dfs(Next, Goal, [Next|Visited], Path).

% Extrae las cabezas de una lista de listas
heads([], []).
heads([[H|_]|T], [H|HT]) :- heads(T, HT).

% Descripción legible de un estado
print_state(State) :-
    State = state(F,W,G,C),
    left_list(state(F,W,G,C), L),
    right_list(state(F,W,G,C), R),
    write('IZQ: '), show_list(L), write(' | DER: '), show_list(R), nl.

left_list(state(F,W,G,C), L4) :-
    L0 = [],
    (F==l -> L1=['Granjero'|L0]; L1=L0),
    (W==l -> L2=['Lobo'|L1]; L2=L1),
    (G==l -> L3=['Cabra'|L2]; L3=L2),
    (C==l -> L4=['Col'|L3]; L4=L3).

right_list(state(F,W,G,C), R4) :-
    R0 = [],
    (F==r -> R1=['Granjero'|R0]; R1=R0),
    (W==r -> R2=['Lobo'|R1]; R2=R1),
    (G==r -> R3=['Cabra'|R2]; R3=R2),
    (C==r -> R4=['Col'|R3]; R4=R3).

show_list([]) :- write('-').
show_list([X]) :- write(X).
show_list([X|Xs]) :- write(X), write(', '), show_list(Xs).

% Describe el movimiento entre dos estados
print_move(state(F,W1,G1,C1), state(_,W2,G2,C2)) :-
    (F==l -> Dir = 'izquierda -> derecha' ; Dir = 'izquierda <- derecha'),
    ( W1 \= W2 -> Name = 'Lobo'
    ; G1 \= G2 -> Name = 'Cabra'
    ; C1 \= C2 -> Name = 'Col'
    ; Name = none
    ),
    ( Name == none -> format('El Granjero cruza solo (~w)~n', [Dir])
    ; format('El Granjero lleva la ~w (~w)~n', [Name, Dir])
    ).

% Imprime la solución paso a paso
print_solution(Path) :-
    Path = [First|_],
    writeln('Solucion (backtracking):'),
    writeln('Paso 0: Estado inicial'),
    print_state(First),
    print_steps(1, Path).

print_steps(_, [_]).
print_steps(N, [Prev,Next|Rest]) :-
    format('~nPaso ~d: ', [N]),
    print_move(Prev, Next),
    print_state(Next),
    N1 is N+1,
    print_steps(N1, [Next|Rest]).

% Punto de entrada (se ejecuta al cargar con swipl -q -s fwgc_solver.pl)
run :-
    ( solve(Path) ->
        print_solution(Path),
        length(Path, L), Steps is L-1,
        format('~nTotal de pasos: ~d~n', [Steps])
    ; writeln('No se encontró solución.')
    ).

File: Deber 2/ejercicio_5/fwgc_solver.py
# Estado como tupla (F, W, G, C): 0 = izquierda, 1 = derecha
START = (0, 0, 0, 0)
GOAL = (1, 1, 1, 1)
NAMES_SHORT = ['F', 'W', 'G', 'C']
NAMES_ES = ['Granjero', 'Lobo', 'Cabra', 'Col']

def is_safe(state):
    F, W, G, C = state
    # Si lobo y cabra se quedan solos en una orilla sin el granjero -> inválido
    if W == G and W != F:
        return False
    # Si cabra y col se quedan solas en una orilla sin el granjero -> inválido
    if G == C and G != F:
        return False
    return True

def neighbors(state):
    F, W, G, C = state
    # Movimientos: granjero solo o con uno que esté en su misma orilla
    candidates = [None, 'W', 'G', 'C']
    idx = {'F': 0, 'W': 1, 'G': 2, 'C': 3}
    for carry in candidates:
        # Debe estar en la misma orilla que el granjero
        if carry is not None:
            if state[idx[carry]] != F:
                continue
        new_state = list(state)
        # Mueve al granjero
        new_state[0] = 1 - new_state[0]
        # Mueve el que lleva (si aplica)
        if carry is not None:
            new_state[idx[carry]] = 1 - new_state[idx[carry]]
        new_state = tuple(new_state)
        if is_safe(new_state):
            yield new_state, carry

# --- Backtracking (DFS) en lugar de BFS ---

def dfs_backtracking(start: tuple = START, goal: tuple = GOAL):
    """Retorna una ruta desde start a goal usando backtracking (DFS).
    Formato de salida: lista de (state, move), comenzando con (start, None).
    Nota: No garantiza mínimos pasos.
    """
    visited = {start}
    path = [(start, None)]
    return _dfs(start, goal, visited, path)

def _dfs(state: tuple, goal: tuple, visited: set, path: list):
    if state == goal:
        return path
    for ns, carry in neighbors(state):
        if ns in visited:
            continue
        visited.add(ns)
        res = _dfs(ns, goal, visited, path + [(ns, carry)])
        if res is not None:
            return res
        visited.remove(ns)
    return None

# --- Salida y utilidades ---

def format_banks(state):
    left = [NAMES_ES[i] for i, side in enumerate(state) if side == 0]
    right = [NAMES_ES[i] for i, side in enumerate(state) if side == 1]
    def fmt(xs): return ', '.join(xs) if xs else '—'
    return f"IZQ: {fmt(left)} | DER: {fmt(right)}"

def describe_move(prev, carry):
    F_prev = prev[0]
    direction = "izquierda -> derecha" if F_prev == 0 else "izquierda <- derecha"
    if carry is None:
        return f"El Granjero cruza solo ({direction})"
    name = {'W': 'Lobo', 'G': 'Cabra', 'C': 'Col'}[carry]
    return f"El Granjero lleva la {name} ({direction})"

def main():
    solution = dfs_backtracking()
    if not solution:
        print("No se encontró solución.")
        return
    print("Solución (backtracking):")
    # Imprime estado inicial
    print(f"Paso 0: Estado inicial")
    print(format_banks(solution[0][0]))
    for i in range(1, len(solution)):
        prev_state = solution[i-1][0]
        state, carry = solution[i]
        print(f"\nPaso {i}: {describe_move(prev_state, carry)}")
        print(format_banks(state))
    print(f"\nTotal de pasos: {len(solution)-1}")

if __name__ == "__main__":
    main()